# Добро пожаловать на разговоры о кино
### В главных ролях:
* @batwomal ака главный по мемам 
* @malkinte 
* @timepota

<div style="text-align: center;">
  <img src="very_important_folder/1.jpeg"/>
</div>

### Перед тем, как узнать насколько глубока кроличья нора, нам нужны все, у кого все еще есть наши методы
<div style="text-align: center;">
  <img src=very_important_folder/2.jpg>
</div>

In [2]:
import movielens_analysis as ma

movies = ma.Movies('../datasets/ml-latest-small/movies.csv')
ratings = ma.Ratings('../datasets/ml-latest-small/ratings.csv')
tags = ma.Tags('../datasets/ml-latest-small/tags.csv')
links = ma.Links('../datasets/ml-latest-small/links.csv')

### Мы погружаемся в дебри тегов:
Данный класс нам наглядно покажет, что интересного люди пишут в тегах, порой там попадаются весьма странные вещи.

* #### Most words (Чем больше слов, тем лучше ~~нет~~)
Я уже говорил, что такое безумие? Безумие - это использовать 16!!! слов в одном теге...

In [2]:
tags.most_words(10)

{'Something for everyone in this one... saw it without and plan on seeing it with kids!': 16,
 'the catholic church is the most corrupt organization in history': 10,
 'villain nonexistent or not needed for good story': 8,
 '06 Oscar Nominated Best Movie - Animation': 7,
 'It was melodramatic and kind of dumb': 7,
 'Oscar (Best Music - Original Score)': 6,
 'stop using useless characters for filler': 6,
 'Oscar (Best Effects - Visual Effects)': 6,
 'Everything you want is here': 5,
 'r:disturbing violent content including rape': 5}

* #### Longest (Самые длиииииииннннные теги)
Ну да, что и следовало ожидать, кто-то решил написать в тегах Войну и Мир.
Ну и насилие, расчлененка и все такое. Спасибо, что заранее об этом предупредили. 

In [3]:
tags.longest(10)

['Something for everyone in this one... saw it without and plan on seeing it with kids!',
 'the catholic church is the most corrupt organization in history',
 'villain nonexistent or not needed for good story',
 'r:disturbing violent content including rape',
 '06 Oscar Nominated Best Movie - Animation',
 'stop using useless characters for filler',
 'Academy award (Best Supporting Actress)',
 'Oscar (Best Effects - Visual Effects)',
 'It was melodramatic and kind of dumb',
 'r:sustained strong stylized violence']

* #### Most words and longest
А вот что будет если мы решим посмотреть самые длинные теги по количеству слов и букв одновременно.
И да, хватит использовать бесполезных персонажей для растягивания хронометража.

In [ ]:
tags.most_words_and_longest(10)

['the catholic church is the most corrupt organization in history',
 'Oscar (Best Effects - Visual Effects)',
 'villain nonexistent or not needed for good story',
 'Something for everyone in this one... saw it without and plan on seeing it with kids!',
 '06 Oscar Nominated Best Movie - Animation',
 'r:disturbing violent content including rape',
 'stop using useless characters for filler',
 'It was melodramatic and kind of dumb']

* #### Most popular (Самые частые теги)
Как и ожидалось, нетфликс и супергероика весьма часто упоминаются в тегах.

In [5]:
tags.most_popular(10)

{'In Netflix queue': 131,
 'atmospheric': 36,
 'superhero': 24,
 'thought-provoking': 24,
 'funny': 23,
 'Disney': 23,
 'surreal': 23,
 'religion': 22,
 'sci-fi': 21,
 'dark comedy': 21}

* #### Tags with (мини поисковик по тегам)
Пожалуй оставлю эти теги без комментариев...

In [6]:
tags.tags_with('poor')

['Poor plot development',
 'Poor story',
 'poor dialogue',
 'poorly paced',
 'rich guy - poor girl']

In [7]:
tags.tags_with('dumb')

['It was melodramatic and kind of dumb', 'dumb']

In [8]:
tags.tags_with('stupid')

['Stupid ending',
 'stupid',
 'stupid but funny',
 'stupid ending',
 'stupid is as stupid does']

In [9]:
tags.tags_with(' ass')

["Harley Quinn's ass", 'bad ass']

* #### Да здравствуют бенчмарки!
<div style="text-align: center;">
  <img src="very_important_folder/6.jpg"/>
</div>

In [3]:
%timeit tags = ma.Tags('../datasets/ml-latest-small/tags.csv')
%timeit tags.most_words(10)
%timeit tags.longest(10)
%timeit tags.most_words_and_longest(10)
%timeit tags.tags_with('poor')

2.58 ms ± 53.9 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
11.2 ms ± 366 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
390 μs ± 23.1 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
11.4 ms ± 644 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
616 μs ± 29.5 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


### А теперь поговорим про фильмы и их жанры:
Интересно в каком году вышло больше всего фильмов? А какой самый частый жанр? Для этого и создан этот класс.

* #### Dist by release (Сколько фильмов вышло за год)
Ничего себе 2002 год продуктивный, аж целых 311 фильмов вышло!

In [11]:
movies.dist_by_release()

{'2002': 311,
 '2006': 295,
 '2001': 294,
 '2007': 284,
 '2000': 283,
 '2009': 282,
 '2003': 279,
 '2004': 279,
 '2014': 278,
 '1996': 276,
 '2015': 274,
 '2005': 273,
 '2008': 269,
 '1999': 263,
 '1997': 260,
 '1995': 259,
 '1998': 258,
 '2011': 254,
 '2010': 247,
 '2013': 239,
 '1994': 237,
 '2012': 233,
 '2016': 218,
 '1993': 198,
 '1992': 167,
 '1988': 165,
 '1987': 153,
 '1990': 147,
 '1991': 147,
 '2017': 147,
 '1989': 142,
 '1986': 139,
 '1985': 126,
 '1984': 101,
 '1981': 92,
 '1980': 89,
 '1982': 87,
 '1983': 83,
 '1979': 69,
 '1977': 63,
 '1973': 59,
 '1978': 59,
 '1965': 47,
 '1971': 47,
 '1974': 45,
 '1976': 44,
 '1964': 43,
 '1967': 42,
 '1968': 42,
 '1975': 42,
 '1966': 42,
 '2018': 41,
 '1962': 40,
 '1972': 39,
 '1963': 39,
 '1959': 37,
 '1960': 37,
 '1955': 36,
 '1969': 35,
 '1961': 34,
 '1970': 33,
 '1957': 33,
 '1958': 31,
 '1953': 30,
 '1956': 30,
 '1940': 25,
 '1949': 25,
 '1954': 23,
 '1942': 23,
 '1939': 23,
 '1946': 23,
 '1951': 22,
 '1950': 21,
 '1947': 20,
 '19

* #### Dist by genres (Самые популярные жанры)
Ну да, драмы и комедии снимают часто, а вот нуар не очень

In [ ]:
movies.dist_by_genres()

{'Drama': 4361,
 'Comedy': 3756,
 'Thriller': 1894,
 'Action': 1828,
 'Romance': 1596,
 'Adventure': 1263,
 'Crime': 1199,
 'Sci-Fi': 980,
 'Horror': 978,
 'Fantasy': 779,
 'Children': 664,
 'Animation': 611,
 'Mystery': 573,
 'Documentary': 440,
 'War': 382,
 'Musical': 334,
 'Western': 167,
 'IMAX': 158,
 'Film-Noir': 87,
 'no genres listed)': 34}

* #### Most genres (Фильмы с наибольшим количеством жанров)
А как они умудрились запихнуть туда аж целых 10 жанров в один фильм?

In [13]:
movies.most_genres(10)

{'Rubber (2010)': 10,
 'Patlabor: The Movie (Kidô keisatsu patorebâ: The Movie) (1989)': 8,
 'Mulan (1998)': 7,
 'Who Framed Roger Rabbit? (1988)': 7,
 'Osmosis Jones (2001)': 7,
 'Interstate 60 (2002)': 7,
 'Robots (2005)': 7,
 'Pulse (2006)': 7,
 'Aqua Teen Hunger Force Colon Movie Film for Theaters (2007)': 7,
 'Enchanted (2007)': 7}

* #### Ну и снова бенчмарки
<div style="text-align: center;">
  <img src="very_important_folder/7.jpg"/>
</div>

In [4]:
%timeit movies = ma.Movies('../datasets/ml-latest-small/movies.csv')
%timeit movies.dist_by_release()
%timeit movies.dist_by_genres()
%timeit movies.most_genres(10)

277 ns ± 27.2 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
19.8 ms ± 674 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
17 ms ± 553 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
460 ns ± 20.7 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


# **Отчет по классу Ratings**

### **Инициализация класса Ratings**

In [18]:
import movielens_analysis as ma
ratings = ma.Ratings('../datasets/ml-latest-small/ratings.csv')

## **Класс Movies**

### **Распределение количества оценок по годам**

In [19]:
ratings.movies.dist_by_year()

{1996: 6040,
 1997: 1916,
 1998: 507,
 1999: 2439,
 2000: 10061,
 2001: 3922,
 2002: 3478,
 2003: 4014,
 2004: 3279,
 2005: 5813,
 2006: 4059,
 2007: 7114,
 2008: 4351,
 2009: 4158,
 2010: 2300,
 2011: 1690,
 2012: 4657,
 2013: 1664,
 2014: 1439,
 2015: 6616,
 2016: 6702,
 2017: 8199,
 2018: 6418}

Как видно из данных выше, наибольшее количество оценок приходится на 2000 год  - 10061, наименьшее на 1998 год - 507.

### **Распределение оценок по количеству их выставления**

In [20]:
ratings.movies.dist_by_rating()

{0.5: 1370,
 1.0: 2811,
 1.5: 1791,
 2.0: 7551,
 2.5: 5550,
 3.0: 20047,
 3.5: 13136,
 4.0: 26818,
 4.5: 8551,
 5.0: 13211}

Как видно из полученных результатов, чаще всего пользователи ставят оценку 4.0 - 26818 раз, реже всего - 0.5 - 1370 раз.

### **Топ-n фильмов по количеству оценок**

In [21]:
ratings.movies.top_by_num_of_ratings(10)

{'Forrest Gump (1994)': 329,
 '"Shawshank Redemption': 317,
 'Pulp Fiction (1994)': 307,
 '"Silence of the Lambs': 279,
 '"Matrix': 278,
 'Star Wars: Episode IV - A New Hope (1977)': 251,
 'Jurassic Park (1993)': 238,
 'Braveheart (1995)': 237,
 'Terminator 2: Judgment Day (1991)': 224,
 "Schindler's List (1993)": 220}

Больше всего оценок получил фильм 'Forrest Gump (1994)' - 329, на 10 месте оказался фильм Schindler's List (1993) -  220 оценок.

### **Топ-n фильмов по среднему рейтингу**

In [22]:
ratings.movies.top_by_ratings(10)

{'The Jinx: The Life and Deaths of Robert Durst (2015)': 5.0,
 'Galaxy of Terror (Quest) (1981)': 5.0,
 'Alien Contamination (1980)': 5.0,
 "I'm the One That I Want (2000)": 5.0,
 'Lesson Faust (1994)': 5.0,
 '"Assignment': 5.0,
 'Mephisto (1981)': 5.0,
 'Black Mirror': 5.0,
 'Dylan Moran: Monster (2004)': 5.0,
 'Bill Hicks: Revelations (1993)': 5.0}

Как видно из данных выше все фильмы, входящие в топ 10, имеют средний рейтинг 5.0.

### **Топ-n фильмов по медиане оценок**

In [23]:
ratings.movies.top_by_ratings(10, metric="median")

{'The Jinx: The Life and Deaths of Robert Durst (2015)': 5.0,
 'Galaxy of Terror (Quest) (1981)': 5.0,
 'Alien Contamination (1980)': 5.0,
 'Troll 2 (1990)': 5.0,
 "I'm the One That I Want (2000)": 5.0,
 '"Chorus Line': 5.0,
 "Guess Who's Coming to Dinner (1967)": 5.0,
 'Children of the Corn IV: The Gathering (1996)': 5.0,
 'Band of Brothers (2001)': 5.0,
 'Lesson Faust (1994)': 5.0}

Как видно из данных выше все фильмы, входящие в топ 10, имеют медиану оценок такую же, как и средний рейтинг - 5.0.

### **Топ-n фильмов по дисперсии оценок**

In [24]:
ratings.movies.top_controversial(10)

{"Ivan's Childhood (a.k.a. My Name is Ivan) (Ivanovo detstvo) (1962)": 5.06,
 'Fanny and Alexander (Fanny och Alexander) (1982)': 5.06,
 'Troll 2 (1990)': 4.5,
 'Lassie (1994)': 4.0,
 '"Zed & Two Noughts': 4.0,
 'Kwaidan (Kaidan) (1964)': 4.0,
 'Emma (2009)': 4.0,
 'Play Time (a.k.a. Playtime) (1967)': 3.72,
 '"Room': 3.56,
 'Peeping Tom (1960)': 3.5}

Дисперсия оценок у фильмов Ivan's Childhood (a.k.a. My Name is Ivan) (Ivanovo detstvo) (1962) и Fanny and Alexander (Fanny och Alexander) (1982) больше 5 и равны 5.06. Так как у этих фильмов сильно разбросаны оценки и их всего 2 - 5 и 0.5. 

### **Топ-n фильмов с наименьшим количеством оценок**

In [25]:
ratings.movies.least_num_of_ratings(10)

{'Best Men (1997)': 1,
 'The Jinx: The Life and Deaths of Robert Durst (2015)': 1,
 'Galaxy of Terror (Quest) (1981)': 1,
 'Looker (1981)': 1,
 'Alien Contamination (1980)': 1,
 'Master of the Flying Guillotine (Du bi quan wang da po xue di zi) (1975)': 1,
 'Nobody Loves Me (Keiner liebt mich) (1994)': 1,
 "I'm the One That I Want (2000)": 1,
 '"Legend of Rita': 1,
 '"Visit': 1}

Выше представлены фильмы были оценены лишь по 1 разу.

### **Фильмы со средним рейтингом выше порогового значения**

In [26]:
ratings.movies.movies_above_threshold(4.5)

{'The Jinx: The Life and Deaths of Robert Durst (2015)': 5.0,
 'Galaxy of Terror (Quest) (1981)': 5.0,
 'Alien Contamination (1980)': 5.0,
 "I'm the One That I Want (2000)": 5.0,
 'Lesson Faust (1994)': 5.0,
 '"Assignment': 5.0,
 'Mephisto (1981)': 5.0,
 'Black Mirror': 5.0,
 'Dylan Moran: Monster (2004)': 5.0,
 'Bill Hicks: Revelations (1993)': 5.0,
 'My Sassy Girl (Yeopgijeogin geunyeo) (2001)': 5.0,
 'Strictly Sexual (2008)': 5.0,
 '"Story of Women (Affaire de femmes': 5.0,
 '"Passenger': 5.0,
 '"Marriage of Maria Braun': 5.0,
 'Lamerica (1994)': 5.0,
 'Entertaining Angels: The Dorothy Day Story (1996)': 5.0,
 'Siam Sunset (1999)': 5.0,
 'Little Murders (1971)': 5.0,
 'Tales of Manhattan (1942)': 5.0,
 '"Calcium Kid': 5.0,
 'Go for Zucker! (Alles auf Zucker!) (2004)': 5.0,
 '"Chump at Oxford': 5.0,
 '"Ugly Duckling and Me!': 5.0,
 '"Breed': 5.0,
 'Sun Alley (Sonnenallee) (1999)': 5.0,
 '"Valet': 5.0,
 'Watching the Detectives (2007)': 5.0,
 'Boy Eats Girl (2005)': 5.0,
 "Garfield's 

Всего фильмов со средним рейтингом выше порогового значения 4.5 - 343.

### **Фильмы с медианой рейтинга выше порогового значения**

In [27]:
ratings.movies.movies_above_threshold(4.5, metric="median")

{'The Jinx: The Life and Deaths of Robert Durst (2015)': 5.0,
 'Galaxy of Terror (Quest) (1981)': 5.0,
 'Alien Contamination (1980)': 5.0,
 'Troll 2 (1990)': 5.0,
 "I'm the One That I Want (2000)": 5.0,
 '"Chorus Line': 5.0,
 "Guess Who's Coming to Dinner (1967)": 5.0,
 'Children of the Corn IV: The Gathering (1996)': 5.0,
 'Band of Brothers (2001)': 5.0,
 'Lesson Faust (1994)': 5.0,
 '"Assignment': 5.0,
 'Mephisto (1981)': 5.0,
 '"Trial': 5.0,
 'Black Mirror: White Christmas (2014)': 5.0,
 'Black Mirror': 5.0,
 'Dylan Moran: Monster (2004)': 5.0,
 'Bill Hicks: Revelations (1993)': 5.0,
 'My Sassy Girl (Yeopgijeogin geunyeo) (2001)': 5.0,
 'Strictly Sexual (2008)': 5.0,
 'Secrets & Lies (1996)': 5.0,
 '"Story of Women (Affaire de femmes': 5.0,
 '"Passenger': 5.0,
 '"Marriage of Maria Braun': 5.0,
 "Man Bites Dog (C'est arrivé près de chez vous) (1992)": 5.0,
 'Lamerica (1994)': 5.0,
 'Entertaining Angels: The Dorothy Day Story (1996)': 5.0,
 'Siam Sunset (1999)': 5.0,
 'Little Murders 

Всего фильмов с медианой рейтинга выше порогового значения 4.5 - 357.

In [3]:
%timeit ratings = ma.Ratings('../datasets/ml-latest-small/ratings.csv')
%timeit ratings.movies.dist_by_year()
%timeit ratings.movies.dist_by_rating()
%timeit ratings.movies.top_by_num_of_ratings(10)
%timeit ratings.movies.top_by_ratings(10)
%timeit ratings.movies.top_by_ratings(10, metric="median")
%timeit ratings.movies.top_controversial(10)
%timeit ratings.movies.least_num_of_ratings(10)
%timeit ratings.movies.movies_above_threshold(4.5)
%timeit ratings.movies.movies_above_threshold(4.5, metric="median")

563 ms ± 19 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
48.6 ms ± 1.3 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
5.75 ms ± 43 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
9.53 ms ± 424 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
31.2 ms ± 1.32 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
33.9 ms ± 1.38 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
38.5 ms ± 727 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
18 ms ± 493 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
31.2 ms ± 445 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
38.1 ms ± 3.74 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


## **Класс Users**

### **Распределение пользователей по количеству оценок**

In [29]:
ratings.users.dist_users_by_num_of_rating()

{414: 2698,
 599: 2478,
 474: 2108,
 448: 1864,
 274: 1346,
 610: 1302,
 68: 1260,
 380: 1218,
 606: 1115,
 288: 1055,
 249: 1046,
 387: 1027,
 182: 977,
 307: 975,
 603: 943,
 298: 939,
 177: 904,
 318: 879,
 232: 862,
 480: 836,
 608: 831,
 600: 763,
 483: 728,
 590: 728,
 105: 722,
 19: 703,
 305: 677,
 489: 648,
 111: 646,
 438: 635,
 217: 613,
 140: 608,
 477: 600,
 555: 578,
 91: 575,
 28: 570,
 219: 528,
 534: 520,
 89: 518,
 64: 517,
 226: 507,
 561: 505,
 18: 502,
 525: 500,
 57: 476,
 381: 474,
 368: 469,
 509: 467,
 469: 465,
 560: 458,
 462: 455,
 292: 446,
 21: 443,
 597: 443,
 42: 440,
 160: 437,
 294: 437,
 580: 436,
 596: 411,
 202: 403,
 275: 403,
 517: 400,
 45: 399,
 156: 398,
 514: 397,
 391: 386,
 567: 385,
 357: 383,
 103: 377,
 339: 371,
 62: 366,
 199: 363,
 125: 360,
 51: 359,
 132: 347,
 66: 345,
 313: 340,
 200: 334,
 221: 331,
 6: 314,
 453: 311,
 50: 310,
 425: 306,
 428: 300,
 573: 299,
 352: 294,
 84: 293,
 122: 292,
 382: 291,
 356: 289,
 135: 279,
 239:

Всего было 610 пользователей, которые оценили фильмы. Больше всего оценок выставил пользователь с id 414 - 2698.

### **Распределение пользователей по средней выставленной оценке**

In [30]:
ratings.users.dist_users_by_rating()

{53: 5.0,
 251: 4.87,
 515: 4.85,
 25: 4.81,
 30: 4.74,
 523: 4.69,
 348: 4.67,
 171: 4.63,
 452: 4.56,
 43: 4.55,
 122: 4.55,
 371: 4.55,
 441: 4.52,
 400: 4.51,
 52: 4.48,
 538: 4.47,
 168: 4.46,
 417: 4.46,
 543: 4.45,
 106: 4.44,
 319: 4.43,
 601: 4.43,
 413: 4.41,
 475: 4.41,
 188: 4.4,
 12: 4.39,
 276: 4.39,
 154: 4.38,
 533: 4.38,
 581: 4.38,
 1: 4.37,
 69: 4.37,
 586: 4.37,
 59: 4.36,
 128: 4.36,
 544: 4.36,
 253: 4.35,
 459: 4.35,
 553: 4.34,
 585: 4.34,
 519: 4.33,
 70: 4.32,
 336: 4.32,
 435: 4.32,
 300: 4.3,
 93: 4.29,
 532: 4.28,
 74: 4.27,
 49: 4.26,
 80: 4.26,
 224: 4.26,
 291: 4.26,
 164: 4.25,
 169: 4.25,
 246: 4.25,
 398: 4.25,
 162: 4.24,
 209: 4.24,
 362: 4.24,
 494: 4.23,
 250: 4.22,
 17: 4.21,
 337: 4.21,
 340: 4.21,
 573: 4.21,
 79: 4.2,
 227: 4.2,
 252: 4.2,
 258: 4.2,
 460: 4.2,
 595: 4.2,
 97: 4.19,
 364: 4.19,
 526: 4.19,
 119: 4.18,
 267: 4.18,
 505: 4.18,
 296: 4.17,
 72: 4.16,
 465: 4.16,
 491: 4.16,
 458: 4.15,
 29: 4.14,
 37: 4.14,
 243: 4.14,
 290: 4.14

Только один пользователь с id 53 ставил всем фильмам оценку - 5.0.

### **Распределение пользователей по медиане выставленных оценок**

In [31]:
ratings.users.dist_users_by_rating(metric="median")

{1: 5.0,
 25: 5.0,
 30: 5.0,
 43: 5.0,
 52: 5.0,
 53: 5.0,
 59: 5.0,
 69: 5.0,
 77: 5.0,
 106: 5.0,
 154: 5.0,
 164: 5.0,
 171: 5.0,
 188: 5.0,
 224: 5.0,
 236: 5.0,
 251: 5.0,
 258: 5.0,
 267: 5.0,
 276: 5.0,
 348: 5.0,
 393: 5.0,
 413: 5.0,
 417: 5.0,
 441: 5.0,
 452: 5.0,
 494: 5.0,
 515: 5.0,
 523: 5.0,
 533: 5.0,
 543: 5.0,
 544: 5.0,
 598: 5.0,
 12: 4.75,
 49: 4.5,
 70: 4.5,
 72: 4.5,
 88: 4.5,
 122: 4.5,
 168: 4.5,
 169: 4.5,
 178: 4.5,
 209: 4.5,
 210: 4.5,
 221: 4.5,
 246: 4.5,
 250: 4.5,
 253: 4.5,
 295: 4.5,
 296: 4.5,
 319: 4.5,
 336: 4.5,
 371: 4.5,
 398: 4.5,
 399: 4.5,
 400: 4.5,
 421: 4.5,
 435: 4.5,
 459: 4.5,
 460: 4.5,
 475: 4.5,
 495: 4.5,
 505: 4.5,
 537: 4.5,
 538: 4.5,
 548: 4.5,
 553: 4.5,
 573: 4.5,
 578: 4.5,
 581: 4.5,
 585: 4.5,
 586: 4.5,
 594: 4.5,
 601: 4.5,
 138: 4.25,
 327: 4.25,
 2: 4.0,
 4: 4.0,
 5: 4.0,
 11: 4.0,
 13: 4.0,
 16: 4.0,
 17: 4.0,
 18: 4.0,
 27: 4.0,
 29: 4.0,
 31: 4.0,
 32: 4.0,
 33: 4.0,
 34: 4.0,
 35: 4.0,
 37: 4.0,
 39: 4.0,
 40: 4.0,

33 пользователя имеют медиану оценок 5.0. То есть они оценивали фильмы на высшую оценку.

### **Топ-n пользователей с наибольшей дисперсией оценок**

In [32]:
ratings.users.top_users_with_biggest_var(10)

{3: 4.26,
 461: 3.1,
 55: 3.09,
 259: 2.94,
 329: 2.92,
 502: 2.76,
 175: 2.75,
 598: 2.7,
 393: 2.61,
 138: 2.44}

Полученные результаты показывают, что у данных пользователей значительная вариативность оценок.

In [33]:
%timeit ratings.users.dist_users_by_num_of_rating()
%timeit ratings.users.dist_users_by_rating()
%timeit ratings.users.dist_users_by_rating(metric="median")
%timeit ratings.users.top_users_with_biggest_var(10)

5.68 ms ± 252 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
7.94 ms ± 254 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
12.7 ms ± 532 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
20.1 ms ± 779 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


# **Отчет по классу Links**

Для последующего анализа данных **MovieLens** нам необходимо инициилизировать класс **_Links()_**, для этого мы создали экземпляр класса Links()- **_links_**, и передали ему путь к файлу с данными MovieLens- **links.csv**, содержащий id номера фильмов базы MovieLens: **movieid**     **imdbid**  **tmdbid**

In [34]:
%timeit ma.Links('../datasets/ml-latest-small/links.csv')
links = ma.Links('../datasets/ml-latest-small/links.csv')
links

6.93 ms ± 365 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


## Общие сведения о классе **Links()**

Создав экземпляр класса Links() мы открыли для себя возможность обработку к базы **MovieLens** по movieid, а по movieid обработать содержимое **imdb**. Кстати, IMDb — Internet Movie Database (IMDb, в переводе с англ. — «Интернет-база кинофильмов») (по-русски ИМДб) — крупнейшая в мире база данных и веб-сайт о кинематографе. В базе собрана информация (по состоянию на конец января 2013 года) о более чем 2 млн 436 тыс.

Все dunder методы и аттрибуты любого класса (методы, начинающиеся и заканчивающиеся двойными нижними подчеркиваниями) - это методы, которые не предназначены для непосредственного пользовательского вызова, но их интерпретация вызывается внутри интерпретатора Python для выполнения определенных действий внутреннего использования в самом классе.
С остальными методами и аттрибутами мы можем взаимодействовать, например мы вызвали аттрибут **'base_url_imdb'**

In [35]:
%timeit links.base_url_imdb
links.base_url_imdb


20.9 ns ± 1.51 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)


'https://www.imdb.com/'

и убедились в том, что базовая ссылка действительно обращается к ресурсу IMDB.

Мы также можем посмотреть и путь к файлу с базой MovieLens, который указали ранее:

In [36]:
%timeit links.filepath
links.filepath

20.3 ns ± 0.569 ns per loop (mean ± std. dev. of 7 runs, 100,000,000 loops each)


'../datasets/ml-latest-small/links.csv'

Дальше мы распарсили сайт imdb с помощью файла links.csv и получили данные по каждому, интересующему нас фильму, передав в метод .get_imdb() список из номеров фильмов - movieid:
<div style="text-align: center;">
  <img src="very_important_folder/8.jpg"/>
</div>

In [37]:
%timeit links.get_imdb([193600, 2, 6, 2467, 903, 2176, 930, 2178])
links.get_imdb([193600, 2, 6, 2467, 903, 2176, 930, 2178])

'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'


7.66 s ± 623 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


'Get movieid 193600 - request failed'


[['movieId',
  'Gross worldwide',
  'imdbId',
  'Title',
  'Runtime',
  'Director',
  'Budget',
  'URL'],
 [2467,
  7153487,
  'tt0091605',
  'The Name of the Rose',
  7800,
  ['Jean-Jacques Annaud'],
  30000000000,
  'https://www.imdb.com/title/tt0091605'],
 [2178,
  4940,
  'tt0068611',
  'Frenzy',
  6960,
  ['Alfred Hitchcock'],
  2000000,
  'https://www.imdb.com/title/tt0068611'],
 [2176,
  10883,
  'tt0040746',
  'Rope',
  4800,
  ['Alfred Hitchcock'],
  1500000,
  'https://www.imdb.com/title/tt0040746'],
 [930,
  117060,
  'tt0038787',
  'Notorious',
  6120,
  ['Alfred Hitchcock'],
  2000000,
  'https://www.imdb.com/title/tt0038787'],
 [903,
  7967183,
  'tt0052357',
  'Vertigo',
  7680,
  ['Alfred Hitchcock'],
  2479000,
  'https://www.imdb.com/title/tt0052357'],
 [6,
  187436818,
  'tt0113277',
  'Heat',
  10200,
  ['Michael Mann'],
  60000000,
  'https://www.imdb.com/title/tt0113277'],
 [2,
  262821940,
  'tt0113497',
  'Jumanji',
  6240,
  ['Joe Johnston'],
  65000000,
  'htt

На выходе мы получили **список списков**- это такой список, элемент которого также является списком! В нашем случае первым элементом списка является заголовок, который отражает название позиции аргумента в последующих списках, а последующие списки - это информация по каждому отдельному фильму, запрошенная через метод .get_imdb().

Количество элементов в основном списке равняется количеству полученных с imdb данных о фильмах. Если какой-либо movieid в links.csv отсутствует, то результат его обработки в итоговый список добавлен не будет.

Давайте передадим в .get_imdb() список полей, которые мы хотим получить, например:

In [38]:
%timeit links.get_imdb([193600, 2, 6, 2467, 903, 2176, 930, 2178], ['Title'])
links.get_imdb([193600, 2, 6, 2467, 903, 2176, 930, 2178], ['Title'])

'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'


8 s ± 736 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


'Get movieid 193600 - request failed'


[['movieId', 'Title'],
 [2467, 'The Name of the Rose'],
 [2178, 'Frenzy'],
 [2176, 'Rope'],
 [930, 'Notorious'],
 [903, 'Vertigo'],
 [6, 'Heat'],
 [2, 'Jumanji']]

Как мы увидели поле movieid добавилось автоматически- это нужно для того, чтобы результат не казался пустым если мы передали неправильное название поля:

In [39]:
%timeit links.get_imdb([193600, 2, 6, 2467, 903, 2176, 930, 2178], ['title'])
links.get_imdb([193600, 2, 6, 2467, 903, 2176, 930, 2178], ['title'])

'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'


8.06 s ± 574 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


'Get movieid 193600 - request failed'


[['movieId', 'title'],
 [2467, None],
 [2178, None],
 [2176, None],
 [930, None],
 [903, None],
 [6, None],
 [2, None]]

Из данного примера в процессе анализа MovieLens мы усволили, что запрашивать поля следует через заглавную букву, например, запросим еще бюджет и ссылку на страничку в интернете:

In [40]:
%timeit links.get_imdb([193600, 2, 6, 2467, 903, 2176, 930, 2178], ['Title', 'Budget', 'URL'])
links.get_imdb([193600, 2, 6, 2467, 903, 2176, 930, 2178], ['Title', 'Budget', 'URL'])

'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'


7.82 s ± 554 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


'Get movieid 193600 - request failed'


[['movieId', 'Title', 'Budget', 'URL'],
 [2467,
  'The Name of the Rose',
  30000000000,
  'https://www.imdb.com/title/tt0091605'],
 [2178, 'Frenzy', 2000000, 'https://www.imdb.com/title/tt0068611'],
 [2176, 'Rope', 1500000, 'https://www.imdb.com/title/tt0040746'],
 [930, 'Notorious', 2000000, 'https://www.imdb.com/title/tt0038787'],
 [903, 'Vertigo', 2479000, 'https://www.imdb.com/title/tt0052357'],
 [6, 'Heat', 60000000, 'https://www.imdb.com/title/tt0113277'],
 [2, 'Jumanji', 65000000, 'https://www.imdb.com/title/tt0113497']]

Отлично! У нас всё отобразилось в нужном порядке. Кстати следует заметить, что бюджет фильма на imdb указан в национальной валюте :\) и в верхней строчке отображен фильм с бюджетом в 30 млрд Итальянских Лир, но об это позже...

Как видно, список списков отсортирован по убыванию movieid и это не случайно, потому как мы в "шапке" можем увидеть насколько глубока кроличья нора...

Чтобы продолжить, давате вызовем метод .get_imdb(), но без указания полей, чтобы получить всю доступную информацию для её дальнейшей обработки, ведь вызывая данный метод мы сохраняем полученные данные в атрибут нашего класса:

In [41]:
%timeit links.get_imdb([193600, 2, 6, 2467, 903, 2176, 930, 2178])
links.get_imdb([193600, 2, 6, 2467, 903, 2176, 930, 2178])

'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'
'Get movieid 193600 - request failed'


8.53 s ± 1.17 s per loop (mean ± std. dev. of 7 runs, 1 loop each)


'Get movieid 193600 - request failed'


[['movieId',
  'Gross worldwide',
  'imdbId',
  'Title',
  'Runtime',
  'Director',
  'Budget',
  'URL'],
 [2467,
  7153487,
  'tt0091605',
  'The Name of the Rose',
  7800,
  ['Jean-Jacques Annaud'],
  30000000000,
  'https://www.imdb.com/title/tt0091605'],
 [2178,
  4940,
  'tt0068611',
  'Frenzy',
  6960,
  ['Alfred Hitchcock'],
  2000000,
  'https://www.imdb.com/title/tt0068611'],
 [2176,
  10883,
  'tt0040746',
  'Rope',
  4800,
  ['Alfred Hitchcock'],
  1500000,
  'https://www.imdb.com/title/tt0040746'],
 [930,
  117060,
  'tt0038787',
  'Notorious',
  6120,
  ['Alfred Hitchcock'],
  2000000,
  'https://www.imdb.com/title/tt0038787'],
 [903,
  7967183,
  'tt0052357',
  'Vertigo',
  7680,
  ['Alfred Hitchcock'],
  2479000,
  'https://www.imdb.com/title/tt0052357'],
 [6,
  187436818,
  'tt0113277',
  'Heat',
  10200,
  ['Michael Mann'],
  60000000,
  'https://www.imdb.com/title/tt0113277'],
 [2,
  262821940,
  'tt0113497',
  'Jumanji',
  6240,
  ['Joe Johnston'],
  65000000,
  'htt

Количество элементов основного списка == количество полученных данных о запрошенных фильмах (за исключением недоступных movieid) минус заголовок.

Результат парсинга мы также можем посмотреть в атрибуте .imdb_content

In [42]:
%timeit links.imdb_content
links.imdb_content

23.5 ns ± 2.35 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)


[['movieId',
  'Gross worldwide',
  'imdbId',
  'Title',
  'Runtime',
  'Director',
  'Budget',
  'URL'],
 [2467,
  7153487,
  'tt0091605',
  'The Name of the Rose',
  7800,
  ['Jean-Jacques Annaud'],
  30000000000,
  'https://www.imdb.com/title/tt0091605'],
 [2178,
  4940,
  'tt0068611',
  'Frenzy',
  6960,
  ['Alfred Hitchcock'],
  2000000,
  'https://www.imdb.com/title/tt0068611'],
 [2176,
  10883,
  'tt0040746',
  'Rope',
  4800,
  ['Alfred Hitchcock'],
  1500000,
  'https://www.imdb.com/title/tt0040746'],
 [930,
  117060,
  'tt0038787',
  'Notorious',
  6120,
  ['Alfred Hitchcock'],
  2000000,
  'https://www.imdb.com/title/tt0038787'],
 [903,
  7967183,
  'tt0052357',
  'Vertigo',
  7680,
  ['Alfred Hitchcock'],
  2479000,
  'https://www.imdb.com/title/tt0052357'],
 [6,
  187436818,
  'tt0113277',
  'Heat',
  10200,
  ['Michael Mann'],
  60000000,
  'https://www.imdb.com/title/tt0113277'],
 [2,
  262821940,
  'tt0113497',
  'Jumanji',
  6240,
  ['Joe Johnston'],
  65000000,
  'htt

Полученные данные, в зависимомости от запрошенных ранее полей, мы обрабатывали при помощи соответствующих методов. Каждый (из указанных ниже) метод возвращает словарь, где ключ- это имя САМОГО-САМОГО, а значение ключа- максимальная сумма достижений ключа, отсортированная по убыванию... Немного о доступных методах:

`.top_directors(top_n)` - отвечает за вывод словаря режиссеров, которые сняли больше всего фильмов из запрошенного списка

`.most_expensive(top_n)` - отвечает за вывод словаря фильмов, бюджет которых (в национальной валюте) был самым дорогим

`.most_profitable(top_n)` - отвечает за вывод словаря фильмов, которые оказались самыми прибыльными (разумеется, в баксах!)

`.longest(top_n)` - отвечает за вывод словаря самых продолжительных фильмов (в секундах)

`.top_cost_per_minute(top_n)` - отвечает за вывод словаря фильмов и во сколько им обошлась стоимость съемки одной минуты
___
*`.top_highest_grossing_directors(top_n)` - отвечает за вывод словаря режиссеров, фильмы которых окупились дороже всего
___
Методы `.get_top_sequence()` и `.isint()`- эти методы являются вспомогательными и они в непосредственной обраотке данных imdb не участвуют, но они оказывают существенное влияние на логику работы скрипта.

+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
## Анализ полученных данных

В аналитике и исследовании данных главное- это объем и количество. Имея больше данных мы имеем место для манёвра оперирования этими данными. В рамках данного исследования ввиду значительных временных затрат на получение больших данных с сайта imdb.com, и в целях их обработки (ну... так тупо интереснее :\) :\) :\) ) мы заранее выгрузили все доступные в links.csv movieid и поместили их в файл **imdb_content.txt** рядом с links.csv

Вызвав метод .get_imdb() без параметров и с доступным файлом imdb_content.txt мы разом получим всю доступную информацию (а это очень много) о фильмах и сможем её обрабатывать с заметными изменениями в результатах обработки

___
### Получение данных .get_imdb() и отображение первых 15 элементов списка

In [43]:
%timeit links.get_imdb()[:10]
links.get_imdb()[:10]

209 ms ± 28.5 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


[['movieId',
  'Gross worldwide',
  'imdbId',
  'Title',
  'Runtime',
  'Director',
  'Budget',
  'URL'],
 [193609, 637327, 'tt0101726', 'Dice Rules', 5280, ['Jay Dubin'], None],
 [193587,
  2373203,
  'tt8391976',
  'Bungou Stray Dogs: Dead Apple',
  5520,
  ['Takuya Igarashi'],
  None],
 [193585, None, 'tt6397426', 'Flint', 5160, ['Bruce Beresford'], None],
 [193583,
  6356284,
  'tt5914996',
  'Nôgêmu nôraifu: Zero',
  6600,
  ['Atsuko Ishizuka'],
  None],
 [193581,
  511132,
  'tt5476944',
  'Kuroshitsuji: Book of the Atlantic',
  6060,
  ['Noriyuki Abe', 'Stephen Hoff'],
  None],
 [193579,
  None,
  'tt5342766',
  'Jon Stewart Has Left the Building',
  3600,
  [None],
  None],
 [193573,
  22069046,
  'tt3837248',
  'Love Live! The School Idol Movie',
  5940,
  ['Takahiko Kyôgoku'],
  None],
 [193571,
  6803548,
  'tt3110014',
  'Gin no saji',
  6660,
  ['Keisuke Yoshida'],
  None],
 [193567,
  109630,
  'tt2323836',
  'Gekijouban Ano hi mita hana no namae wo bokutachi wa mada shir

Как видим - movieid имеет обратный links.csv порядок

___
### Аналитика ТОП-10 режиссеров - .top_directors()

In [44]:
%timeit links.top_directors(10)
links.top_directors(10)

4.05 ms ± 740 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


{'Woody Allen': 48,
 'Alfred Hitchcock': 37,
 'Steven Spielberg': 32,
 'Clint Eastwood': 32,
 'Martin Scorsese': 30,
 'Ron Howard': 24,
 'Steven Soderbergh': 24,
 'Ridley Scott': 24,
 'Spike Lee': 21,
 'Barry Levinson': 21}

Вуди Аллен за свою карьеру (с 1965 года) и до момента создания базы MovieLens (2018) год отснял 48 фильмов - это почти 1 фильм в год!

В тройку лидеров вошли также Альфред Хичкок и знаменитый Стивен Спилберг.

___
### Аналитика ТОП-10 прибыльных режиссеров - .top_highest_grossing_directors()

In [45]:
%timeit links.top_highest_grossing_directors(10)
links.top_highest_grossing_directors(10)

8.17 ms ± 437 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


{'James Cameron': 2923710708,
 'J.J. Abrams': 2071310218,
 'Colin Trevorrow': 1671537444,
 'Joss Whedon': 1405018048,
 'Rian Johnson': 1334407706,
 'J.A. Bayona': 1310469037,
 'Chris Buck': 1306450154,
 'Jennifer Lee': 1306450154,
 'Bill Condon': 1266115964,
 'Brad Bird': 1243225667}

Хо-Хо! Какую информацию мы уже успели получить? Мы узнали, что режессировать много фильмов не означает, что это качественные фильмы, которые хочет смотреть зритель! Из **ТОП-10 режиссеров** в **ТОП-10 прибыльных режиссеров** не вошел ни один "скорострел"...

В **ТОП-10 прибыльных режиссеров** лидирующие позиции заняли Джеймс Кэмерон с его Титаником, Аватаром, Терминаторами... тут не поспоришь! Второе место у Бати Звёздных Войн - Дж.Дж. Абрамса. Третье место по праву принадлежит Колину Треворрову с его серией фильмов Парк Юрского периода...

___
### Аналитика ТОП-10 прибыльных фильмов - .most_profitable()

In [46]:
%timeit links.most_profitable(10)
links.most_profitable(10)

8.13 ms ± 1.58 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)


{'Avatar': 2686710708,
 'Titanic': 2064812968,
 'Star Wars: Episode VII - The Force Awakens': 1826310218,
 'Avengers: Infinity War': 1731415039,
 'Jurassic World': 1521537444,
 'Fast & Furious 7': 1325342457,
 'Harry Potter and the Deathly Hallows - Part 2': 1217499744,
 'Frozen': 1156450154,
 'Avengers: Age of Ultron': 1155018048,
 'Black Panther': 1149926083}

А вот и подтверждение предыдущему факту - Аватар (Джеймс Кэмерон) собрал 2,68 млрд $, Карл, сделав при этом режиссера самым прибыльным режиссером и заняв первую строчку данного ТОП-10! До этого фильма Кэмерон отснял фильм-легенду-катастрофу - Титаник (второе место).
Третье место, конечно-же - Звёздные войны с их бомбическим эпизодом!

___
### Аналитика ТОП-10 дорогих фильмов - .most_expensive()

In [47]:
%timeit links.most_expensive(10)
links.most_expensive(10)

6.62 ms ± 1.13 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)


{'The Name of the Rose': 30000000000,
 'La vita è bella': 15000000000,
 'Gwoemul': 12215500000,
 'Busanhaeng': 10000000000,
 'Ah-ga-ssi': 10000000000,
 'Fantaghirò': 5000000000,
 'Kaguya-hime no monogatari': 5000000000,
 'Rupan sansei: Kariosutoro no shiro': 5000000000,
 'Chinjeolhan geumjassi': 4200000000,
 'Il gattopardo': 2900000000}

Аналитика требует вдумчивого и осознанного подхода. Согласно данным результатам - самый дорогой фильм это 'The Name of the Rose', но здесь кроется ошибка! Данный фильм был снят в конце 80-х годов в Италии, где, на тот момент, национальной валютой являлась Итальянская Лира, вот и получается, что на съемки данного фильма ушло 30 млрд Лир! В данном ТОПе имеют место выбросы, т.к. мы не можем под один общий знаменатель подводить фильмы, бюджет которых был в разных валютах...

Повторим вызов метода, увеличив ТОП до ТОП-35:

In [48]:
%timeit links.most_expensive(35)
links.most_expensive(35)

5.78 ms ± 292 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


{'The Name of the Rose': 30000000000,
 'La vita è bella': 15000000000,
 'Gwoemul': 12215500000,
 'Busanhaeng': 10000000000,
 'Ah-ga-ssi': 10000000000,
 'Fantaghirò': 5000000000,
 'Kaguya-hime no monogatari': 5000000000,
 'Rupan sansei: Kariosutoro no shiro': 5000000000,
 'Chinjeolhan geumjassi': 4200000000,
 'Il gattopardo': 2900000000,
 'Mononoke-hime': 2400000000,
 'Innocence': 2000000000,
 'Yôkai daisensô': 1300000000,
 'Akira': 1100000000,
 'Majo no takkyûbin': 800000000,
 'Ôritsu uchûgun Oneamisu no tsubasa': 800000000,
 'Tango': 700000000,
 'Lola Montès': 650000000,
 'Chi bi': 553632000,
 '3 Idiots': 550000000,
 'Zombi 2': 410000000,
 'Suriyothai': 400000000,
 'Zashchitniki': 380000000,
 'Kimi no na wa.': 370000000,
 'Abre los ojos': 370000000,
 'Kôkaku kidôtai: Stand Alone Complex Solid State Society': 360000000,
 'Kaidan': 350000000,
 'Kôkaku Kidôtai': 330000000,
 'Astérix & Obélix : Mission Cléopâtre': 327000000,
 'Avengers: Infinity War': 321000000,
 'Star Wars: Episode VIII 

Вот здесь, начиная с 30й строчки и кроется САМЫЙ ДОРОГОЙ ФИЛЬМ на момент 2018 года - это 'Avengers: Infinity War' с его космобюджетом в 321000000$. CGI - дорогущее удовольствие, однако...

Всё, что попало в ТОП-35 до этого фильма- фильмы, бюджет которых исчислялся в национальной валюте, что нас мало интересует, потому как рекорды мировых сборов регистрируются в вечнозеленых билетах банка дядюшки Сэма и логичнее сделать конвертацию для достоверности текущего ТОПа.

___
### Аналитика ТОП-10 долгих фильмов - .longest()

In [49]:
%timeit links.longest(10)
links.longest(10)

5.09 ms ± 131 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


{'The Godfather Trilogy: 1901-1980': 34980,
 'Tiexi qu': 33060,
 'O.J.: Made in America': 28020,
 'La meglio gioventù': 22440,
 'Novecento': 19020,
 'As I Was Moving Ahead Occasionally I Saw Brief Glimpses of Beauty': 17280,
 'Bis ans Ende der Welt': 17220,
 'Human': 15780,
 'The Greatest Story Ever Told': 15600,
 'Gettysburg': 15240}

Фильм, он же кинофильм- может быть разделен на серии, части, эпизоды и всё такое прочее, и при этом являться цельной картиной! В наш ТОП-10 долгих фильмов попал Крёстный отец и... ТВ документальные фильмы! Как неожиданно... 

___
### Аналитика ТОП-10 дорогих фильмов за минуту съемки - .top_cost_per_minute()

In [50]:
%timeit links.top_cost_per_minute(10)
links.top_cost_per_minute(10)

13.5 ms ± 3.19 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)


{'The Name of the Rose': 230769230.77,
 'La vita è bella': 129310344.83,
 'Gwoemul': 101795833.33,
 'Busanhaeng': 84745762.71,
 'Ah-ga-ssi': 68965517.24,
 'Rupan sansei: Kariosutoro no shiro': 50000000.0,
 'Chinjeolhan geumjassi': 36521739.13,
 'Kaguya-hime no monogatari': 36496350.36,
 'Fantaghirò': 25000000.0,
 'Innocence': 20000000.0}

И снова туда-же- прямиком в ловушку национальной валюты. Поднимем ставки до ТОП-35!

In [51]:
%timeit links.top_cost_per_minute(35)
links.top_cost_per_minute(35)

10.9 ms ± 948 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


{'The Name of the Rose': 230769230.77,
 'La vita è bella': 129310344.83,
 'Gwoemul': 101795833.33,
 'Busanhaeng': 84745762.71,
 'Ah-ga-ssi': 68965517.24,
 'Rupan sansei: Kariosutoro no shiro': 50000000.0,
 'Chinjeolhan geumjassi': 36521739.13,
 'Kaguya-hime no monogatari': 36496350.36,
 'Fantaghirò': 25000000.0,
 'Innocence': 20000000.0,
 'Mononoke-hime': 18045112.78,
 'Il gattopardo': 15591397.85,
 'Yôkai daisensô': 10483870.97,
 'Akira': 8870967.74,
 'Majo no takkyûbin': 7766990.29,
 'Ôritsu uchûgun Oneamisu no tsubasa': 6611570.25,
 'Tango': 6086956.52,
 'Lola Montès': 5603448.28,
 'Afro Samurai': 4599230.77,
 'Zombi 2': 4505494.51,
 'Zashchitniki': 4269662.92,
 'Kôkaku Kidôtai': 3975903.61,
 'Chi bi': 3818151.72,
 'The Pacific': 3773584.91,
 'Kimi no na wa.': 3490566.04,
 'Kôkaku kidôtai: Stand Alone Complex Solid State Society': 3333333.33,
 'Paprika': 3333333.33,
 '3 Idiots': 3235294.12,
 'Abre los ojos': 3109243.7,
 'Astérix & Obélix : Mission Cléopâtre': 3056074.77,
 'T2 3-D: B

По данным результатам сложно оценить объективность полученной информации. В данных обстоятельствах логичнее добавить поле с валютой и провести конвертацию, потому как текущая объективность результатов крайне сомнительна :)

Отчет подготовил mlkinte- пир и студент основной формы обучения Школы-21 (Сбербанк)

Чукча- не писатель

### Вам наверное инетерсно, а все ли тут правильно? Да, все протестировано!
<div style="text-align: center;">
  <img src="very_important_folder/3.jpg"/>
  <img src="very_important_folder/4.jpg"/>
</div>

## На этом все, всем спасибо за внимание!
<div style="text-align: center;">
  <img src="very_important_folder/5.jpg"/>
</div>